ads/google/page-level-experiments.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isExperimentOn, toggleExperiment} from '../../src/experiments';
+
+/**
+ * @typedef {{
+ *   isTrafficEligible: !function(!Window):boolean,
+ *   branches: !Array<string>
+ * }}
+ */
+export let ExperimentInfo;
+
+
+/**
+ * In some browser implementations of Math.random(), sequential calls of
+ * Math.random() are correlated and can cause a bias.  In particular,
+ * if the previous random() call was < 0.001 (as it will be if we select
+ * into an experiment), the next value could be less than 0.5 more than
+ * 50.7% of the time.  This provides an implementation that roots down into
+ * the crypto API, when available, to produce less biased samples.
+ *
+ * @return {number} Pseudo-random floating-point value on the range [0, 1).
+ */
+function slowButAccuratePrng() {
+  // TODO(tdrl): Implement.
+  return Math.random();
+}
+
+/**
+ * Container for alternate random number generator implementations.  This
+ * allows us to set an "accurate" PRNG for branch selection, but to mock it
+ * out easily in tests.
+ *
+ * @visibleForTesting
+ * @const {!{accuratePrng: function():number}}
+ */
+export const RANDOM_NUMBER_GENERATORS = {
+  accuratePrng: slowButAccuratePrng,
+};
+
+/**
+ * Selects, uniformly at random, a single item from the array.
+ * @param {!Array<string>} arr Object to select from.
+ * @return {?string} Single item from arr or null if arr was empty.
+ */
+function selectRandomItem(arr) {
+  const rn = RANDOM_NUMBER_GENERATORS.accuratePrng();
+  return arr[Math.floor(rn * arr.length)] || null;
+}
+
+/**
+ * Selects which page-level experiments, if any, a given amp-ad will
+ * participate in.  If a given experiment name is already set (including to
+ * the null / no branches selected state), this won't alter its state.
+ *
+ * Check whether a given experiment is set using isExperimentOn(win,
+ * experimentName) and, if it is on, look for which branch is selected in
+ * win.pageExperimentBranches[experimentName].
+ *
+ * @param {!Window} win Window context on which to save experiment
+ *     selection state.
+ * @param {!Object<string,!ExperimentInfo>} experiments  Set of experiments to
+ *     configure for this page load.
+ * @visibleForTesting
+ */
+export function randomlySelectUnsetPageExperiments(win, experiments) {
+  win.pageExperimentBranches = win.pageExperimentBranches || {};
```

Comment:
It's page-level, rather than element-level info, so I needed some page shared state to track.  (Ensures that multiple slots on the same page don't reach different conclusions about their experiment status.)  Is there a better place to store such shared state?